src/stash.c

```diff
@@ -541,50 +665,87 @@ int git_stash_save(
 	const char *message,
 	uint32_t flags)
 {
-	git_index *index = NULL;
+	git_stash_save_options opts = GIT_STASH_SAVE_OPTIONS_INIT;
+	opts.stasher = stasher;
+	opts.message = message;
+	opts.flags = flags;
+	return git_stash_save_with_opts(out, repo, &opts);
+}
+
+int git_stash_save_with_opts(
+	git_oid *out, git_repository *repo, git_stash_save_options *opts)
+{
+	git_index *index = NULL, *paths_index = NULL;
 	git_commit *b_commit = NULL, *i_commit = NULL, *u_commit = NULL;
 	git_buf msg = GIT_BUF_INIT;
+	git_tree *tree = NULL;
+	git_reference *head = NULL;
 	int error;
 
 	GIT_ASSERT_ARG(out);
 	GIT_ASSERT_ARG(repo);
-	GIT_ASSERT_ARG(stasher);
 
```

Comment:
Since it appears `opts.stasher` is required, should we instead do something like:
```c
GIT_ASSERT_ARG(opts);
GIT_ASSERT_ARG(opts->stasher); // not sure if this is the best way to validate properties
```

Reply:
It seems `GIT_ASSERT_ARG` is used when NULL is not allowed for the argument, and checking for NULL and providing a default value is used when NULL is allowed. So I agree we can use `GIT_ASSERT_ARG(opts)` inside `git_stash_save_with_opts`.

The problem is when the user calls `git_stash_save`, GIT_STASH_SAVE_OPTIONS_INIT initializes `stasher` with NULL, in which case if we don't allow `stasher` to be NULL inside `git_stash_save_with_opts` it won't work.

I think a possible solution could be `GIT_ASSERT_ARG(stasher)` inside `git_stash_save`, and `GIT_ASSERT_ARG(opts && opts->stasher)` inside `git_stash_save_with_opts